src/textengine.rs

```diff
@@ -1,104 +1,132 @@
-use super::annotation::Annotation;
-use super::cas::Cas;
-use super::engine;
-use super::error::PipelineError;
+use std::path::{Path, PathBuf};
+
+use lazy_static::lazy_static;
 use regex::Regex;
-use std::fs;
-use std::path::PathBuf;
 use walkdir::WalkDir;
 
+use crate::annotation::Annotation;
+use crate::cas::Cas;
+use crate::engine::{Engine, Reader};
+use crate::error::{PipelineError, Result};
+
+#[derive(Debug)]
 pub struct PrintEngine {
-    pub annotation: String,
+    annotation: String,
 }
-pub struct SentenceEngine();
-pub struct Tokenizer();
+
+#[derive(Debug, Default)]
+pub struct SentenceEngine(());
+
+#[derive(Debug, Default)]
+pub struct Tokenizer(());
+
+#[derive(Debug)]
 pub struct RegexEngine {
-    pub pattern: Regex,
-    pub annotation: String,
+    pattern: Regex,
+    annotation: String,
 }
+
+#[derive(Debug)]
 pub struct SimpleDocumentReader {
-    pub documents: Vec<PathBuf>,
-    pub input_dir: String,
-    pub document_index: u32,
-    pub documents_len: u32,
+    documents: std::vec::IntoIter<PathBuf>,
 }
 
-impl engine::Engine for SentenceEngine {
-    fn process(&self, cas: &mut Cas) -> Result<(), PipelineError> {
-        lazy_static! {
-            static ref RE: Regex = Regex::new(r"[^.!?]*[.!?]").unwrap();
-        }
-        let mut annotations:Vec<Annotation> =  Vec::new();
-        for cap in RE.find_iter(cas.text.as_str()) {
-            let begin = cap.start() as i32;
-            let end = cap.end() as i32;
-            let annot = Annotation { begin, end };
-            annotations.push(annot);
+impl PrintEngine {
+    pub fn new(annotation: String) -> Self {
+        Self { annotation }
+    }
+}
+
+impl SentenceEngine {
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+impl Tokenizer {
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+impl RegexEngine {
+    pub fn new(pattern: Regex, annotation: String) -> Self {
+        Self {
+            pattern,
+            annotation,
         }
-        cas.insert_annotations("sentence", annotations);
-        Ok(())
     }
 }
-impl engine::Engine for Tokenizer {
-    fn process(&self, cas: &mut Cas) -> Result<(), PipelineError> {
+
+impl Engine for SentenceEngine {
+    fn process(&self, cas: &mut Cas) -> Result<()> {
         lazy_static! {
-            static ref RE: Regex = Regex::new(r"\w+").unwrap();
-        }
-        let mut annotations:Vec<Annotation> =  Vec::new();
-        for cap in RE.find_iter(cas.text.as_str()) {
-            let begin = cap.start() as i32;
-            let end = cap.end() as i32;
-            let annot = Annotation { begin, end };
-            annotations.push(annot);
+            static ref RE: RegexEngine =
+                RegexEngine::new(Regex::new(r"[^.!?]*[.!?]").unwrap(), "sentence".into());
         }
-        cas.insert_annotations("token", annotations);
-        Ok(())
+
+        RE.process(cas)
```

Comment:
Since you're using a regex anyway, we can offload the work here to the regex engine rather than copy/pasting the implementation.

Reply:
I actually wanted to do this, but didnt know how to create the Engines being type engine, using the self::default is what I was missing to improve this part of the code, thank you very much for the changes